lib/galaxy/webapps/galaxy/api/tools.py

```diff
@@ -426,6 +428,51 @@ def citations(self, trans: GalaxyWebTransaction, id, **kwds):
             rval.append(citation.to_dict('bibtex'))
         return rval
 
+    @expose_api
+    def conversion(self, trans: GalaxyWebTransaction, tool_id, payload, **kwd):
+        converter = self._get_tool(tool_id, user=trans.user)
+        target_type = payload.get("target_type")
+        source_type = payload.get("source_type")
+        input_src = payload.get("src")
+        input_id = payload.get("id")
+        # List of string of dependencies
+        try:
+            deps = trans.app.datatypes_registry.converter_deps[source_type][target_type]
+        except KeyError:
+            deps = {}
+        # Generate parameter dictionary
+        params = {}
+        # determine input parameter name and add to params
+        input_name = 'input1'
+        for key, value in converter.inputs.items():
+            if deps and value.name in deps:
+                params[value.name] = deps[value.name]
+            elif value.type == 'data':
+                input_name = key
+
+        params[input_name] = {
+            "values": [
+                {
+                    "id": input_id,
+                    "src": input_src,
+                }
+            ],
+            "batch": input_src == "hdca",
+        }
+
+        history_id = payload.get('history_id')
+        if history_id:
+            decoded_id = self.decode_id(history_id)
+        else:
+            encoded_id = int(trans.app.security.decode_id(input_id))
+            decoded_id = self.hda_manager.get_accessible(encoded_id, trans.user).history_id
+
+        target_history = self.history_manager.get_owned(decoded_id, trans.user, current_history=trans.history)
```

Comment:
@jmchilton suggested a test case that does not have a history id which prompted this. If I remove that test case, I can make history_id required.

Reply:
``history_id`` shouldn't be required - you can drop the ``current_history=trans.history`` if you'd like though? It will only break checking owned histories for the anonymous user I believe.

We're not defaulting to that history, we're defaulting to the history the target dataset is in and I think that is the right thing to do. The caller shouldn't need to know where that HDA is IMO.

Reply:
We want to kill trans.history, right ? The frontend needs to decide what the target history is, so you can work in multiple tabs and not have bizarre results.

Reply:
> It will only break checking owned histories for the anonymous user I believe.

Why wouldn't this work for anoymous users ? Their histories still have ids, right ?

Reply:
> Why wouldn't this work for anoymous users ? Their histories still have ids, right ?

I'm nervous we're talking about different things here.

``current_history=trans.history`` is used in the ``is_owner`` check - it isn't serving as some sort of default here - it is required to do the security check for anonymous users.

``
    def is_owner(self, history, user, current_history=None, **kwargs):
        """
        True if the current user is the owner of the given history.
        """
        # anon users are only allowed to view their current history
        if self.user_manager.is_anonymous(user):
            if current_history and history == current_history:
                return True
            return False
        return super().is_owner(history, user)
```

I'm sure anonymous histories have an ID but I don't think they have a user right? I think we need ``trans.history`` in order to verify that user belongs to that history.

>  The frontend needs to decide what the target history is, so you can work in multiple tabs and not have bizarre results.

I think the frontend deciding to always send a history_id is different than the deciding the API should accept inputs without specifying a ``history_id``. We should not be designing the API endpoints for the client - the client should use API endpoints that make sense. It makes sense that this allow you to pick a history, but the HDAs history is an obvious default and should be allowed by the API to make the API better.

Reply:
True, I was slightly confused here, and I missed the history_id being parsed out of the payload .... one more thing that's nice about fastAPI, that's harder to miss. The client's not passing the history_id currently, or did I miss that (https://github.com/galaxyproject/galaxy/pull/11799/files#r704566777) ?